src/visor/blockdb/unspent.go

```diff
@@ -37,368 +38,363 @@ func (e ErrUnspentNotExist) Error() string {
 	}
 
-	return &unspentMeta{
-		Bucket: *bkt,
-	}, nil
+	return &out, nil
 }
 
-func (m unspentMeta) getXorHashWithTx(tx *bolt.Tx) (cipher.SHA256, error) {
-	if v := m.GetWithTx(tx, xorhashKey); v != nil {
-		var hash cipher.SHA256
-		copy(hash[:], v[:])
-		return hash, nil
+func (pl pool) getAll(tx *dbutil.Tx) (coin.UxArray, error) {
+	var uxa coin.UxArray
+
+	if err := dbutil.ForEach(tx, UnspentPoolBkt, func(_, v []byte) error {
+		var ux coin.UxOut
+		if err := encoder.DeserializeRaw(v, &ux); err != nil {
+			return err
+		}
+
+		uxa = append(uxa, ux)
+		return nil
+	}); err != nil {
+		return nil, err
 	}
 
-	return cipher.SHA256{}, nil
+	return uxa, nil
 }
 
-func (m *unspentMeta) setXorHashWithTx(tx *bolt.Tx, hash cipher.SHA256) error {
-	return m.PutWithTx(tx, xorhashKey, hash[:])
+func (pl pool) set(tx *dbutil.Tx, hash cipher.SHA256, ux coin.UxOut) error {
+	return dbutil.PutBucketValue(tx, UnspentPoolBkt, hash[:], encoder.Serialize(ux))
 }
 
-type pool struct {
-	bucket.Bucket
+func (pl *pool) delete(tx *dbutil.Tx, hash cipher.SHA256) error {
+	return dbutil.Delete(tx, UnspentPoolBkt, hash[:])
 }
 
-func newPool(db *bolt.DB) (*pool, error) {
-	bkt, err := bucket.New(unspentPoolBkt, db)
-	if err != nil {
+type poolAddrIndex struct{}
+
+func (p poolAddrIndex) get(tx *dbutil.Tx, addr cipher.Address) ([]cipher.SHA256, error) {
+	var hashes []cipher.SHA256
+
+	if ok, err := dbutil.GetBucketObjectDecoded(tx, UnspentPoolAddrIndexBkt, addr.Bytes(), &hashes); err != nil {
 		return nil, err
+	} else if !ok {
+		return nil, nil
 	}
 
-	return &pool{
-		Bucket: *bkt,
-	}, nil
+	return hashes, nil
 }
 
-func (pl pool) getWithTx(tx *bolt.Tx, hash cipher.SHA256) (*coin.UxOut, bool, error) {
-	if v := pl.GetWithTx(tx, hash[:]); v != nil {
-		var out coin.UxOut
-		if err := encoder.DeserializeRaw(v, &out); err != nil {
-			return nil, false, err
+func (p poolAddrIndex) set(tx *dbutil.Tx, addr cipher.Address, hashes []cipher.SHA256) error {
+	hashesMap := make(map[cipher.SHA256]struct{}, len(hashes))
+	for _, h := range hashes {
+		if _, ok := hashesMap[h]; ok {
+			return errors.New("poolAddrIndex.set: hashes array contains duplicate")
 		}
-		return &out, true, nil
+
+		hashesMap[h] = struct{}{}
 	}
-	return nil, false, nil
-}
 
-func (pl pool) setWithTx(tx *bolt.Tx, hash cipher.SHA256, ux coin.UxOut) error {
-	v := encoder.Serialize(ux)
-	return pl.PutWithTx(tx, hash[:], v)
+	encodedHashes := encoder.Serialize(hashes)
+	return dbutil.PutBucketValue(tx, UnspentPoolAddrIndexBkt, addr.Bytes(), encodedHashes)
 }
 
-func (pl *pool) deleteWithTx(tx *bolt.Tx, hash cipher.SHA256) error {
-	return pl.DeleteWithTx(tx, hash[:])
-}
+// adjust adds and removes hashes from an address -> hashes index
+// TODO -- if necessary, this can be optimized further to accept multiple addresses at once,
+// so that all get queries can be performed before the set
+func (p poolAddrIndex) adjust(tx *dbutil.Tx, addr cipher.Address, addHashes, rmHashes []cipher.SHA256) error {
```

Comment:
Because adding and removing happens at the same time.  If you separate it, it takes two reads and two writes instead of one read and one write (and interleaved reads and writes are slow).  Its not a public method, only called two places so the argument position should not be wrong since tests would catch this.